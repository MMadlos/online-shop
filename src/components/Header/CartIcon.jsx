import { FontAwesomeIcon } from "@fortawesome/react-fontawesome"
import { faCartShopping } from "@fortawesome/free-solid-svg-icons"

function CartIcon({ cartQuantity }) {
	const cartHasProduct = cartQuantity !== 0

	return (
		<div className="icon-cart">
			{cartHasProduct && (
				<div className="cart-counter">
					<p>{cartQuantity}</p>
				</div>
			)}

			<FontAwesomeIcon icon={faCartShopping} />
		</div>
	)
}

export default CartIcon
