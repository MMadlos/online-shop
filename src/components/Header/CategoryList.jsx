import { Link } from "react-router-dom"
import { useContext } from "react"
import { ShopContext } from "../../App"
import { getCategories } from "../../utilities/utilities"

function CategoryList() {
	const { productList, selectedCategory, setSelectedCategory } = useContext(ShopContext)
	const categoryList = ["All products", ...getCategories(productList)]
	const handleClickCategory = (e) => setSelectedCategory(e.target.textContent)

	return (
		<div className="category-list">
			<ul>
				{categoryList.map((category, index) => {
					const isCategorySelected = selectedCategory.toLowerCase() === category.toLowerCase()
					const addClassIfSelected = isCategorySelected ? " selected" : ""

					return (
						<li
							key={index}
							className={"category" + addClassIfSelected}>
							<Link
								to="/"
								onClick={handleClickCategory}>
								{category}
							</Link>
						</li>
					)
				})}
			</ul>
		</div>
	)
}

export default CategoryList
