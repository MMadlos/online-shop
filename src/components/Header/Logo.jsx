import { Link } from "react-router-dom"
import { ShopContext } from "../../App"
import { useContext } from "react"

function Logo() {
	const { setSelectedCategory } = useContext(ShopContext)

	return (
		<h1>
			<Link
				to="/"
				onClick={() => setSelectedCategory("All products")}>
				OSHOP{" "}
			</Link>
		</h1>
	)
}

export default Logo
