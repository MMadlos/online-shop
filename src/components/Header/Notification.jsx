function Notification() {
	return (
		<div className="notification-add">
			<p>Added to your cart</p>
		</div>
	)
}

export default Notification
