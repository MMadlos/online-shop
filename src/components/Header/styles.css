/* HEADER */

header {
	width: 100%;
	background: var(--black);
	box-shadow: 4px 4px 15px 0px rgba(0, 0, 0, 0.05);
	color: var(--white);
	padding: 0;

	@media (min-width: 1024px) {
		padding: 16px 10vw;
	}

	position: sticky;
	top: 0;

	.header-container {
		width: 100%;
		height: 100%;
		display: flex;
		justify-content: space-between;
		align-items: center;

		.fa-bars {
			font-size: 24px;
			padding: 16px;

			@media (min-width: 1024px) {
				display: none;
			}
		}

		> h1 {
			font-size: 18px;
			font-style: italic;
			font-weight: 900;
			line-height: normal;
			letter-spacing: 0.9px;
			text-transform: uppercase;

			@media (min-width: 1024px) {
				font-size: 24px;
				letter-spacing: 1.2px;
			}

			> a {
				color: inherit;
			}
		}

		.category-list {
			display: none;

			@media (min-width: 1024px) {
				display: block;

				> ul {
					display: flex;
					gap: 24px;

					.category {
						font-size: 14px;
						font-weight: 700;
						text-transform: uppercase;
						cursor: pointer;

						&.selected {
							color: var(--accent-yellow);
							cursor: default;
						}

						&:not(.selected):hover {
							color: var(--accent-yellow);
						}

						> a {
							color: inherit;
							cursor: inherit;
						}
					}
				}
			}
		}

		.icon-cart {
			color: var(--white);
			position: relative;
			margin-right: 16px;

			@media (min-width: 1024px) {
				margin-right: 0;

				> svg:hover,
				.selected {
					color: var(--accent-yellow);
				}
			}

			.cart-counter {
				height: 20px;
				width: 20px;
				padding: 4px;

				background-color: var(--accent-yellow);
				border-radius: 50%;
				border: 2px solid var(--black);

				position: absolute;
				right: 0;
				transform: translateX(50%) translateY(-50%);

				display: flex;
				justify-content: center;
				align-items: center;

				> p {
					color: var(--black);
					font-size: 10px;
					font-weight: 800;
				}
			}

			.fa-cart-shopping {
				font-size: 25px;
			}
		}

		.notification-add {
			background-color: var(--accent-yellow);
			padding: 8px 24px;
			font-size: 12px;
			font-weight: 600;
			color: var(--black);
			border-radius: 2px;

			position: absolute;
			right: 16px;

			animation: wobble-horizontal-top 0.8s linear both;

			@media (min-width: 1024px) {
				position: absolute;
				top: -16px;
				right: 0;
				margin-right: 10vw;

				padding: 16px 32px;
				font-size: 16px;
			}
		}
	}
}

@keyframes wobble-horizontal-top {
	0%,
	100% {
		transform: translate(0, 125%);
		transform-origin: 50% 50%;
	}
	15% {
		transform: translate(-16px, 125%) rotate(6deg);
	}
	30% {
		transform: translate(8px, 125%) rotate(-6deg);
	}
	45% {
		transform: translate(-8px, 125%) rotate(3.6deg);
	}
	60% {
		transform: translate(4px, 125%) rotate(-2.4deg);
	}
	75% {
		transform: translate(-6px, 125%) rotate(1.2deg);
	}
}

/* CATEGORY PAGE - Menu that opens in mobile version*/

#root:has(.categories) {
	height: 100vh;
	overflow: hidden;
}

.categories {
	@media (min-width: 1024px) {
		display: none;
	}

	position: absolute;
	top: 0;
	left: 0;

	width: 100vw;
	height: 100vh;
	z-index: 9;

	padding: 40px 24px;
	background: var(--accent-yellow);
	background: radial-gradient(circle at top left, var(--accent-yellow) 25%, var(--white) 100%);

	.title {
		display: flex;
		justify-content: space-between;
		align-items: center;
		color: #72778d;

		> p {
			font-size: 12px;
			font-weight: 700;
			text-transform: uppercase;
		}

		> svg {
			padding: 12px;
			font-size: 1rem;
		}
	}

	.separator {
		width: 100%;
		background-color: #72778d;
		height: 1px;
	}

	.category-list.mobile {
		display: block;

		> ul {
			display: flex;
			flex-direction: column;
			gap: 32px;
			margin-top: 40px;
		}

		.category {
			color: var(--black);
			font-size: 24px;
			font-weight: 700;
			text-transform: uppercase;

			&.selected {
				color: var(--accent-purple);
			}

			> a {
				color: inherit;
			}
		}
	}
}
