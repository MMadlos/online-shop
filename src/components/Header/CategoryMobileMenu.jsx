import { FontAwesomeIcon } from "@fortawesome/react-fontawesome"
import { faChevronLeft } from "@fortawesome/free-solid-svg-icons"
import CategoryList from "./CategoryList"

function CategoryMobileMenu({ onClickClose, onClickCategory }) {
	return (
		<nav className="categories mobile">
			<div className="title">
				<p>Categories</p>
				<FontAwesomeIcon
					icon={faChevronLeft}
					onClick={onClickClose}
				/>
			</div>
			<div className="separator"></div>

			<CategoryList
				isMobile={true}
				onClickCategory={onClickCategory}
			/>
		</nav>
	)
}

export default CategoryMobileMenu
