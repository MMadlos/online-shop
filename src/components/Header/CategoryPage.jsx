import { FontAwesomeIcon } from "@fortawesome/react-fontawesome"
import { faChevronLeft } from "@fortawesome/free-solid-svg-icons"
import CategoryList from "./CategoryList"

function CategoryPage({ onClickClose, onClickCategory }) {
	return (
		<nav className="categories mobile">
			<div className="title">
				<p>Categories</p>
				<FontAwesomeIcon
					icon={faChevronLeft}
					onClick={onClickClose}
				/>
			</div>
			<div className="separator"></div>

			<CategoryList
				styles="mobile"
				onClickCategory={onClickCategory}
			/>
		</nav>
	)
}

export default CategoryPage
