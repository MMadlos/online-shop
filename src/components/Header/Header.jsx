import "./styles.css"

import { useState } from "react"
import { Link } from "react-router-dom"

import { FontAwesomeIcon } from "@fortawesome/react-fontawesome"
import { faBars } from "@fortawesome/free-solid-svg-icons"

import CartIcon from "./CartIcon"
import CategoryList from "./CategoryList"
import CategoryPage from "./CategoryPage"
import Notification from "./Notification"

function HeaderContainer({ children }) {
	return (
		<header>
			<div className="header-container">{children}</div>
		</header>
	)
}

function Header({ cartQuantity }) {
	const [isNavOpen, setIsNavOpen] = useState(false)

	const [currentCartQuantity, setCurrentCartQuantity] = useState(cartQuantity)
	const [isNotificationOpen, setIsNotificationOpen] = useState(false)

	if (cartQuantity < currentCartQuantity) {
		setCurrentCartQuantity(cartQuantity)
		return
	}

	if (cartQuantity > currentCartQuantity) {
		setIsNotificationOpen(true)
		setCurrentCartQuantity(cartQuantity)

		setTimeout(setIsNotificationOpen, 2500, false)
	}

	return (
		<HeaderContainer>
			{isNavOpen && (
				<CategoryPage
					onClickClose={() => setIsNavOpen(false)}
					onClickCategory={() => setIsNavOpen(false)}
				/>
			)}
			<FontAwesomeIcon
				icon={faBars}
				onClick={() => setIsNavOpen(true)}
			/>

			<h1>
				<Link to="/">OSHOP </Link>
			</h1>

			<CategoryList />

			<Link to="/cart">
				<CartIcon cartQuantity={cartQuantity} />
			</Link>
			{isNotificationOpen && <Notification />}
		</HeaderContainer>
	)
}

export default Header
