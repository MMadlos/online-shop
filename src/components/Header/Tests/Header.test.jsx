import { describe, expect, it, vi } from "vitest"
import { render, screen } from "@testing-library/react"
import Header from "../Header"

import { BrowserRouter } from "react-router-dom"
import userEvent from "@testing-library/user-event"
import { ShopContext } from "../../../App"

vi.mock("../CategoryMobileMenu", () => ({
	default: () => {
		return <div data-testid="category-mobile-menu"></div>
	},
}))

vi.mock("../CategoryList", () => ({
	default: () => {
		return <div data-testid="category-list"></div>
	},
}))

vi.mock("../Notification", () => ({
	default: () => {
		return <div data-testid="notification"></div>
	},
}))

describe("Header", () => {
	it.skip("renders component", () => {
		const { container } = render(
			<BrowserRouter>
				<Header />
			</BrowserRouter>
		)

		expect(container).toMatchSnapshot()
	})

	describe("Mobile menu icon behaviour", () => {
		it("renders icon faBars with class mobile", () => {
			render(
				<BrowserRouter>
					<Header />
				</BrowserRouter>
			)

			const svgEl = screen.getByTestId("icon-menu-mobile")

			expect(svgEl).toHaveClass("fa-bars")
			expect(svgEl).toHaveClass("mobile")
		})

		it("opens mobile menu if faBars icon is clicked", async () => {
			const user = userEvent.setup()

			render(
				<BrowserRouter>
					<Header />
				</BrowserRouter>
			)

			const svgEl = screen.getByTestId("icon-menu-mobile")

			await user.click(svgEl)

			const categoryMobileMenu = screen.getByTestId("category-mobile-menu")
			expect(categoryMobileMenu).toBeInTheDocument()
		})

		it("doesn't open mobile menu if faBars is not clicked", () => {
			render(
				<BrowserRouter>
					<Header />
				</BrowserRouter>
			)

			const categoryMobileMenu = screen.queryByTestId("category-mobile-menu")
			expect(categoryMobileMenu).not.toBeInTheDocument()
		})
	})

	describe("Notification behaviour", () => {
		it("renders Notification when a product is added to Cart", () => {
			const mockCurrentCartList = [
				{
					id: 0,
					name: "Product 1",
				},
			]

			const mockNewCartList = [
				{
					id: 0,
					name: "Product 1",
				},
				{
					id: 1,
					name: "Product 2",
				},
			]

			const { rerender } = render(
				<BrowserRouter>
					<ShopContext.Provider value={{ cartList: mockCurrentCartList }}>
						<Header />
					</ShopContext.Provider>
				</BrowserRouter>
			)

			rerender(
				<BrowserRouter>
					<ShopContext.Provider value={{ cartList: mockNewCartList }}>
						<Header />
					</ShopContext.Provider>
				</BrowserRouter>
			)

			const notification = screen.getByTestId("notification")
			expect(notification).toBeInTheDocument()
		})

		it("doesn't render Notification when a product is removed from Cart", () => {
			const mockCurrentCartList = [
				{
					id: 0,
					name: "Product 1",
				},
				{
					id: 1,
					name: "Product 2",
				},
			]

			const mockNewCartList = [
				{
					id: 0,
					name: "Product 1",
				},
			]

			const { rerender } = render(
				<BrowserRouter>
					<ShopContext.Provider value={{ cartList: mockCurrentCartList }}>
						<Header />
					</ShopContext.Provider>
				</BrowserRouter>
			)

			rerender(
				<BrowserRouter>
					<ShopContext.Provider value={{ cartList: mockNewCartList }}>
						<Header />
					</ShopContext.Provider>
				</BrowserRouter>
			)

			const notification = screen.queryByTestId("notification")
			expect(notification).not.toBeInTheDocument()
		})

		it("doesn't render Notification when cartList doesn't change", () => {
			const mockCurrentCartList = [
				{
					id: 0,
					name: "Product 1",
				},
				{
					id: 1,
					name: "Product 2",
				},
			]

			const mockNewCartList = [
				{
					id: 0,
					name: "Product 1",
				},
				{
					id: 1,
					name: "Product 2",
				},
			]

			const { rerender } = render(
				<BrowserRouter>
					<ShopContext.Provider value={{ cartList: mockCurrentCartList }}>
						<Header />
					</ShopContext.Provider>
				</BrowserRouter>
			)

			rerender(
				<BrowserRouter>
					<ShopContext.Provider value={{ cartList: mockNewCartList }}>
						<Header />
					</ShopContext.Provider>
				</BrowserRouter>
			)

			const notification = screen.queryByTestId("notification")
			expect(notification).not.toBeInTheDocument()
		})

		it("doesn't render Notification when cartList is empty", () => {
			render(
				<BrowserRouter>
					<ShopContext.Provider value={{ cartList: [] }}>
						<Header />
					</ShopContext.Provider>
				</BrowserRouter>
			)

			const notification = screen.queryByTestId("notification")
			expect(notification).not.toBeInTheDocument()
		})

		it("doesn't render Notification when cartList remains empty", () => {
			const { rerender } = render(
				<BrowserRouter>
					<ShopContext.Provider value={{ cartList: [] }}>
						<Header />
					</ShopContext.Provider>
				</BrowserRouter>
			)

			rerender(
				<BrowserRouter>
					<ShopContext.Provider value={{ cartList: [] }}>
						<Header />
					</ShopContext.Provider>
				</BrowserRouter>
			)

			const notification = screen.queryByTestId("notification")
			expect(notification).not.toBeInTheDocument()
		})
	})

	describe("CartIcon behaviour", () => {
		it("renders quantity number of cart items when is > 0", () => {
			const mockCartList = [
				{
					id: 0,
					name: "Product 1",
				},
			]

			const cartQuantity = mockCartList.length

			render(
				<BrowserRouter>
					<ShopContext.Provider value={{ cartList: mockCartList }}>
						<Header />
					</ShopContext.Provider>
				</BrowserRouter>
			)

			const cartNumberElement = screen.queryByText(cartQuantity)
			expect(cartNumberElement).toBeInTheDocument()
		})
		it("doesn't render quantity number of cart items when is = 0", () => {
			const mockCartList = []

			const cartQuantity = mockCartList.length
			render(
				<BrowserRouter>
					<ShopContext.Provider value={{ cartList: mockCartList }}>
						<Header />
					</ShopContext.Provider>
				</BrowserRouter>
			)

			const cartNumberElement = screen.queryByText(cartQuantity)
			expect(cartNumberElement).not.toBeInTheDocument()
		})
	})
})
