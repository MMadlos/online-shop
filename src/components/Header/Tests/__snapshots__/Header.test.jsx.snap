// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Header > renders component 1`] = `
<div>
  <header>
    <div
      class="header-container"
    >
      <svg
        aria-hidden="true"
        class="svg-inline--fa fa-bars "
        data-icon="bars"
        data-prefix="fas"
        focusable="false"
        role="img"
        viewBox="0 0 448 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M0 96C0 78.3 14.3 64 32 64H416c17.7 0 32 14.3 32 32s-14.3 32-32 32H32C14.3 128 0 113.7 0 96zM0 256c0-17.7 14.3-32 32-32H416c17.7 0 32 14.3 32 32s-14.3 32-32 32H32c-17.7 0-32-14.3-32-32zM448 416c0 17.7-14.3 32-32 32H32c-17.7 0-32-14.3-32-32s14.3-32 32-32H416c17.7 0 32 14.3 32 32z"
          fill="currentColor"
        />
      </svg>
      <h1>
        <a
          href="/"
        >
          OSHOP
           
        </a>
      </h1>
      <div
        class="category-list"
      >
        <ul>
          <li
            class="category selected"
          >
            <a
              href="/"
            >
              All products
            </a>
          </li>
        </ul>
      </div>
      <a
        href="/cart"
      >
        <div
          class="icon-cart"
        >
          <svg
            aria-hidden="true"
            class="svg-inline--fa fa-cart-shopping "
            data-icon="cart-shopping"
            data-prefix="fas"
            focusable="false"
            role="img"
            viewBox="0 0 576 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M0 24C0 10.7 10.7 0 24 0H69.5c22 0 41.5 12.8 50.6 32h411c26.3 0 45.5 25 38.6 50.4l-41 152.3c-8.5 31.4-37 53.3-69.5 53.3H170.7l5.4 28.5c2.2 11.3 12.1 19.5 23.6 19.5H488c13.3 0 24 10.7 24 24s-10.7 24-24 24H199.7c-34.6 0-64.3-24.6-70.7-58.5L77.4 54.5c-.7-3.8-4-6.5-7.9-6.5H24C10.7 48 0 37.3 0 24zM128 464a48 48 0 1 1 96 0 48 48 0 1 1 -96 0zm336-48a48 48 0 1 1 0 96 48 48 0 1 1 0-96z"
              fill="currentColor"
            />
          </svg>
        </div>
      </a>
    </div>
  </header>
</div>
`;
